src/service-worker/core.js

```diff
@@ -0,0 +1,309 @@
+            // Don't prune the file-version we just added.
+            if (cursor.primaryKey === key) {
+              return '';
+            }
+
+            // We'll want to remove this from the cache, too.
+            const version = cursor.value.version;
+            cursor.delete();
+            return version;
+          });
+        });
+      }).then(removedVersions => {
+        removedVersions.forEach(version => {
+          // Is the return value for the file-version we added (we explicitly
+          // returned an empty string for it)?
+          if (!version) {
+            return;
+          }
+
+          const deleteUrl = versionedUrl(url, version);
+          // We only want to delete RTV files.
+          if (url !== deleteUrl) {
+            cache.delete(deleteUrl);
+          }
+        });
+      });
+    }
+
+    return response;
+  });
+}
+
+
+/**
+ * Gets the version we have cached for this file. It's either:
+ *  - The requestVersion, meaning we have this explicit version cached.
+ *  - Some older version
+ *  - An empty string, meaning we have nothing cached for this file.
+ *
+ * @param {string} requestFile
+ * @param {string} requestVersion
+ * @return {!Promise<string>} Not really a promise, but a Promise-like.
+ */
+function getCachedVersion(requestFile, requestVersion) {
+  return db.transaction('js-files', 'readonly').run(transaction => {
+    const files = transaction.objectStore('js-files');
+    const id = [requestFile, requestVersion];
+
+    // First check if we have this exact file-version.
+    return files.index('fileVersions').get(id).then(file => {
+      // We have it cached!
+      if (file) {
+        return file.version;
+      }
+
+      // Do we have any versions of this file cached?
+      // To get the first cached file, we must open a cursor. We won't actually
+      // iterate past the first match, though, since we do not advance the
+      // cursor.
+      return files.index('files').openCursor(requestFile).iterate(cursor => {
+        return cursor.value.version;
+      }).then(versions => {
+        // We have a version in cache (cause we iterated over it)!
+        if (versions.length) {
+          return versions[0];
+        }
+
+        // This is the first request we've seen for this file.
+        return '';
+      });
+    });
+  });
+}
+
+self.addEventListener('install', install => {
+  install.waitUntil(Promise.all([cachePromise, dbPromise]));
+});
+
+// Setup the Fetch listener
+// My assumptions:
+//   - Doc requests one uniform AMP release version for all files, anything
+//     else is malarkey.
+//   - The requested version is always the newest AMP version.
+self.addEventListener('fetch', event => {
+  const request = event.request;
+  const clientId = event.clientId;
+  const url = request.url;
+
+  // We only cache CDN JS files, and we need a clientId to do our magic.
+  if (!clientId || !isCdnJsFile(url)) {
+    event.respondWith(fetch(request));
```

Comment:
I'd recommend leaving out the `event.respondWith(fetch(request));` and just `return`ing here. That will accomplish the same thing if there aren't any other `fetch` event handlers, but if you do register more `fetch` event handlers at some point in the future, it means that they will get their own chance to respond to the event.

(See https://googlechrome.github.io/samples/service-worker/multiple-handlers/index.html)